sherpa/optmethods/optscipy.py

```diff
@@ -0,0 +1,498 @@
+from sherpa.optmethods.buildin import OptMethod
+from sherpa.utils.types import (ArrayType,
+                                OptFunc, OptReturn,
+                                StatFunc)
+from sherpa.models.parameter import hugeval
+
+# Will be filled out programmatically at the bottom of this file
+__all__ = []
+
+
+def convert_bounds_to_scipy(parmins: ArrayType,
+                            parmaxes: ArrayType,
+                            requires_finite_bounds: bool) -> list[tuple[float | None, float | None]] | None:
+    """Convert parmins and parmaxes to format used by `scipy.optimize` for bounds.
+
+    Parameters
+    ----------
+    parmins : ArrayType
+        The minimum values for the parameters.
+    parmaxes : ArrayType
+        The maximum values for the parameters.
+    requires_finite_bounds : bool
+        If True, the scipy function requires finite bounds.
+
+    Returns
+    -------
+    list[tuple[float | None, float | None]] | None
+        The bounds in the format used by scipy.optimize, or None if
+        there are no bounds.
+    """
+    if np.allclose(parmins, -hugeval) and np.allclose(parmaxes, hugeval):
+        bounds = None
+    else:
+        bounds = [(None if pmin == -hugeval else pmin,
+                 None if pmax == hugeval else pmax)
+                for pmin, pmax in zip(parmins, parmaxes)]
+    if requires_finite_bounds:
+        if None in np.array(bounds):
+            raise ValueError("The optimizer requires finite bounds, but "
+                             "the Sherpa model has some bounds set to HUGEVAL.")
+    return bounds
+
+
+# At first sight, this looks like it should be a decorator, but instead
+# we make it a higher-order function, because decorated functions cannot
+# be pickled.
+# https://pythonicthoughtssnippets.github.io/2020/08/09/PTS13-rethinking-python-decorators.html
+def wrap_scipy_fcn(func: Callable,
+                   requires_finite_bounds: bool,
+                   stat: StatFunc,
+                   x0: ArrayType,
+                   xmin: ArrayType,
+                   xmax: ArrayType,
+                   **kwargs) -> OptReturn:
+    """Wrap a function in scipy.optimize to the Sherpa interface.
+    """
+    sig = inspect.signature(func)
+
+    def stat_wrapper(x):
+        # The function is called with the parameters
+        # and returns the statistic and per-bin values
+        return stat(x)[0]
+
+    converted_args: dict = {}
+    if 'x0' in sig.parameters.keys():
+        converted_args['x0'] = x0
+    if 'bounds' in sig.parameters.keys():
+        converted_args['bounds'] = convert_bounds_to_scipy(xmin, xmax,
+                                                            requires_finite_bounds)
+
+    result = func(stat_wrapper, **converted_args, **kwargs)
+    for arg in ['bounds', 'ranges']:
+        if arg in converted_args:
+            result[f'input_{arg}'] = converted_args[arg]
+    return (result.success, result.x, result.fun, result.message, result)
```

Comment:
I commented on this before but it looks like it has disappeared during development.

`converted_args` can only contain `x0` and `bounds` keys, so why does line 118 include a check for `ranges`? Should it also be looking in `kwargs` too (as that's the only way I can see that `ranges` can be set)?